### PR TITLE
Core: Implement Device Mapping & GPU SMMU

### PR DESCRIPTION
**Pull request by FernandoS27**
Memory in Tegra X1 works very differently to how it's currently emulated. Normally there's physical memory of (4Gb), a virtual memory address space used by processes (applications like games for example) and device virtual memory spaces used by peripherals (GPU, DSP, bluetooth, etc). This device memory space is normally called (SMMU) System Memory Management Unit or IOMMU. With this address space, the different devices can map over physical memory used by many different applications to one they can only see. This has the advantage that it allows sharing memory among many multiple processes in the OS with each device.

This PR implements a general simplified version of the SMMU and implements it for the GPU. With some differences, first, we use full 34 bits of the AS instead of just 32 bits for pinning memory, second, we use a common SMMU page file for every process in the GPU instead of switching it per channel.

Currently in yuzu., the GPU uses a memory model of GMMU (GPU's MMU) -> PMMU (the main application's virtual space) -> physical memory to handle memory overall. This is great when there is only one application using the GPU at a time and it's never shared with other processes. 95% of the time this is what happens in the switch. However, it is possible to have multiple processes running and rendering concurrently as is the case of overlays and applets like inline software keyboard. It's also possible to have an application suspended while another is running (an applet is running while the game is suspended). With the implementation of the SMMU, it is now possible to share the GPU resources without any issues whatsoever.

**Advantages of new SMMU:**

- Not perfect but more accurate than before.
- Uses less memory overall for tracking GPU memory.
- More optimizations possible, in the future.
- Multiprocess use of the emulated GPU.
- Other devices can use the device mapper in case it's needed or be extended through correct implementation of KDevicePageTable.
- Total memory usage of the SMMU is about 80Mb counting tracking. The tracking alone on none-SMMU version costed 256MBs.

**Disadvantages of the new SMMU.**
- Harder tracking of resources.
- Complicated.
- More accurate than before but not perfectly accurate to avoid sacrificing performance.

Current issues:

- [X] Pikmin 4 seems to get some vertex explosions after transitioning in many worlds. (Fixed)
- [X] ~~There's some memory leaking after closing an application.~~ (Unrelated, the leak is on master)
- [X] It's not working with 6Gb/8Gb memory layouts. (fixed)
- [ ] Needs more cleanup.

### CODE REVIEW

File: src/core/hle/service/nvdrv/core/nvmap.cpp [hunk 1]
```diff
@@ -236,6 +259,11 @@ std::optional<NvMap::FreeInfo> NvMap::FreeHandle(Handle::Id handle, bool interna
                     std::scoped_lock queueLock(unmap_queue_lock);
                     UnmapHandle(*handle_description);
                 }
+                if (handle_description->allocated) {
```
Comment (bylaws) [inline src/core/hle/service/nvdrv/core/nvmap.cpp]: This isn't correct to do, why exactly is it required?

File: src/core/hle/service/nvdrv/devices/nvmap.cpp [hunk 2]
```diff
@@ -116,15 +123,15 @@ NvResult nvmap::IocAlloc(IocAllocParams& params) {
         return NvResult::InsufficientMemory;
     }
 
-    const auto result =
-        handle_description->Alloc(params.flags, params.align, params.kind, params.address);
+    const auto result = file.AllocateHandle(params.handle, params.flags, params.align, params.kind,
```
Comment (bylaws) [inline src/core/hle/service/nvdrv/devices/nvmap.cpp]: The alloc ioctl isn't supposed to map memory onto the smmu here like this does, why is it done here?

**Review by bylaws — CHANGES_REQUESTED** [review]

File: src/core/hle/service/nvdrv/devices/nvmap.cpp (hunk 2 above)
Comment (FernandoS27) [reply src/core/hle/service/nvdrv/devices/nvmap.cpp]: oh that was reverted at some point.

File: src/core/hle/service/nvdrv/core/nvmap.cpp (hunk 1 above)
Comment (FernandoS27) [reply src/core/hle/service/nvdrv/core/nvmap.cpp]: I think you are looking at an old commit.

**Comment by kkxiegouquan** [comment]
l hope to test this feature

**Comment by FernandoS27** [comment]
> l hope to test this feature

this alone means almost nothing to users but it will allow a LOT of cool stuffs soon.

**Review by kkxiegouquan — COMMENTED** [review]
This feature should be great!!

**Comment by Slexer** [comment]
How much work is left for a complete UMA implementation? Is SMMU a big chunk of it or just a little bit?

**Comment by FernandoS27** [comment]
> How much work is left for a complete UMA implementation? Is SMMU a big chunk of it or just a little bit?

For a real good and accurate UMA implementation; a lot of work. SMMU makes it easier for making the mirrors needed for DMI.

After that We'll do only DMI for downloads, which will improve things a bit.

Doing full DMI will be a challenge if we want to keep or even increase the compatibility we have. It will take a lot of time. I can't give an ETA.

**Comment by brujo5** [comment]
by DMI do they refer to direct memory import?  If yes, I remember that in Skyline they said that it only worked perfectly on a7xx and on a6xx you had to restart the phone or the games start to crash and it was a hardware bug, is that true?

**Comment by FernandoS27** [comment]
> by DMI do they refer to direct memory import? If yes, I remember that in Skyline they said that it only worked perfectly on a7xx and on a6xx you had to restart the phone or the games start to crash and it was a hardware bug, is that true?

Yes. It's true.

**Review by kkxiegouquan — APPROVED** [review]

**Comment by Hariu7** [comment]
How much fps it will boost

File: src/core/hle/service/nvdrv/devices/nvhost_vic.cpp [hunk 3]
```diff
@@ -68,14 +68,20 @@ NvResult nvhost_vic::Ioctl3(DeviceFD fd, Ioctl command, std::span<const u8> inpu
...
 void nvhost_vic::OnClose(DeviceFD fd) {
     auto& host1x_file = core.Host1xDeviceFile();
     const auto iter = host1x_file.fd_to_id.find(fd);
     if (iter != host1x_file.fd_to_id.end()) {
         system.GPU().ClearCdmaInstance(iter->second);
     }
+    auto it = sessions.find(fd);
+    if (it != sessions.end()) {
+        sessions.erase(it);
+    }
```
Comment (liamwhite) [inline src/core/hle/service/nvdrv/devices/nvhost_vic.cpp:84]: [sessions.erase(fd)](https://en.cppreference.com/w/cpp/container/unordered_map/erase)

File: src/core/hle/service/nvdrv/nvdrv.cpp [hunk 4]
```diff
@@ -45,13 +45,22 @@ void EventInterface::FreeEvent(Kernel::KEvent* event) {
...
-    server_manager->RegisterNamedService("nvdrv:t",
-                                         std::make_shared<NVDRV>(system, module, "nvdrv:t"));
+    const auto NvdrvInterfaceFactoryForApplication = [&, module] {
+        return std::make_shared<NVDRV>(system, module, "nvdrv");
+    };
+    const auto NvdrvInterfaceFactoryForApplets = [&, module] {
+        return std::make_shared<NVDRV>(system, module, "nvdrv:a");
+    };
+    const auto NvdrvInterfaceFactoryForSysmodules = [&, module] {
+        return std::make_shared<NVDRV>(system, module, "nvdrv:a");
```
Comment (liamwhite) [inline src/core/hle/service/nvdrv/nvdrv.cpp:55]: nvdrv:s

File: src/core/hle/service/nvdrv/nvdrv.cpp [hunk 5]
```diff
@@ -45,13 +45,22 @@ void EventInterface::FreeEvent(Kernel::KEvent* event) {
...
+    };
+    const auto NvdrvInterfaceFactory = [&, module] {
```
Comment (liamwhite) [inline src/core/hle/service/nvdrv/nvdrv.cpp:57]: NvDrvInterfaceFactoryForTesting

File: src/core/device_memory_manager.inc [hunk 6]
```diff
@@ -313,6 +353,18 @@ void DeviceMemoryManager<Traits>::UnregisterProcess(size_t id) {
...
+        if (!locked) {
+            counter_guard.lock();
+            locked = true;
+        }
+    };
+    SCOPE_EXIT({
+        if (locked) {
+            counter_guard.unlock();
+        }
+    });
```
Comment (liamwhite) [inline src/core/device_memory_manager.inc:523]:
```cpp
std::unique_lock<std::mutex> lk(counter_guard, std::defer_lock);
const auto Lock = [&] {
    if (!lk) {
        lk.lock();
    }
};
```
and then the scope exit part is handled by the unique_lock so it's not needed to write it

File: src/core/hle/service/nvdrv/core/container.cpp [hunk 7]
```diff
@@ -48,16 +57,65 @@ size_t Container::OpenSession(Kernel::KProcess* process) {
...
+
+            cur_addr = next_address;
+        }
+        session.has_preallocated_area = false;
+        auto start_region = (region_size >> 15) >= 1024 ? smmu.Allocate(region_size) : 0;
+        if (start_region != 0) {
+            session.mapper = std::make_unique<HeapMapper>(region_start, start_region, region_size,
+                                                          smmu_id, impl->host1x);
+            session.has_preallocated_area = true;
+            LOG_CRITICAL(Debug, "Preallocation created!");
```
Comment (liamwhite) [inline src/core/hle/service/nvdrv/core/container.cpp:106]: Unnecessary critical log

File: src/core/device_memory_manager.inc [hunk 8]
```diff
@@ -251,6 +255,47 @@ void DeviceMemoryManager<Traits>::Unmap(DAddr address, size_t size) {
...
 }
+template <typename Traits>
+void DeviceMemoryManager<Traits>::TrackContinuityImpl(DAddr address, VAddr virtual_address,
+                                                      size_t size, size_t process_id) {
+    Core::Memory::Memory* process_memory = registered_processes[process_id];
+    size_t start_page_d = address >> Memory::YUZU_PAGEBITS;
+    size_t num_pages = Common::AlignUp(size, Memory::YUZU_PAGESIZE) >> Memory::YUZU_PAGEBITS;
+    uintptr_t last_ptr = 0;
+    size_t page_count = 1;
+    for (size_t i = num_pages; i > 0; i--) {
```
Comment (liamwhite) [inline src/core/device_memory_manager.inc:286]: Wouldn't it make more sense to do this as a forward loop? Also it seems you are off by one...

**Review by liamwhite — APPROVED** [review]

**Pull request merged**